### PR TITLE
[Settings v2] Manage Social Accounts

### PR DESCRIPTION
**Pull request by IniZio**
ref DEV-1860
ref DEV-2165

[CI](https://github.com/IniZio/authgear-server/tree/DEV-1860-settings-oauth-1)

Dependencies
- [x] https://github.com/authgear/authgear-server/pull/4792

https://github.com/user-attachments/assets/bd534467-8e58-4374-8523-0adcb4b57c86

### CODE REVIEW

File: pkg/auth/handler/webapp/sso_callback.go [hunk 1]
```diff
@@ -38,6 +41,22 @@ func (h *SSOCallbackHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
 
+	s := session.GetSession(r.Context())
+
+	if state.AccountManagementToken != "" {
+		_, err = h.AccountManagement.FinishAddingIdentityOAuth(s, &accountmanagement.FinishAddingIdentityOAuthInput{
+			Token: state.AccountManagementToken,
+			Query: r.URL.Query().Encode(),
+		})
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusBadRequest)
```
Comment (louischan-oursky) [inline pkg/auth/handler/webapp/sso_callback.go:52]: We should handle the error properly.

1. The error could be an API error. For example, linking to a OAuth account that is already associated with an existing Authgear user. This should be handled by showing the error in the original route where the add operation is triggered.
2. The error could be an OAuth error produced by the OAuth provider, such as complaining the client secret is invalid. Let's see how is this error handled in v1.
3. The error could be an I/O error. This error can be panic to let the panic middleware to handle.

**Review by louischan-oursky — CHANGES_REQUESTED** [review]

File: pkg/auth/handler/webapp/sso_callback.go (hunk 1 above)
Comment (IniZio) [reply pkg/auth/handler/webapp/sso_callback.go]: Updated to use authflow like error handling

*API error e.g. conflict identity*
Screen Shot 2024-10-14 at 19 42 45

*Unknown error e.g. secret error*
Screen Shot 2024-10-14 at 19 42 38

File: resources/authgear/templates/en/web/authflowv2/__error.html [hunk 2]
```diff
@@ -156,8 +156,23 @@
         <span>{{ include "v2.error.developer-email-required" nil }}</span>
       {{ end }}
     {{ else if eq .Error.reason "AccountManagementDuplicatedIdentity" }}
-      {{ $dict := (dict "IdentityTypeIncoming" "other") }}
-      <span>{{ include "v2.error.account-conflict" $dict }}</span>
+      {{ $info := $.Error.info }}
+      {{ $dict :=
+        (dict
+          "FlowType" $info.FlowType
```
Comment (louischan-oursky) [inline resources/authgear/templates/en/web/authflowv2/__error.html:162]: There is no FlowType. Remove this.

File: pkg/auth/handler/webapp/sso_callback.go [hunk 3]
```diff
@@ -44,16 +47,25 @@ func (h *SSOCallbackHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
+
 		_, err = h.AccountManagement.FinishAddingIdentityOAuth(s, &accountmanagement.FinishAddingIdentityOAuthInput{
 			Token: state.AccountManagementToken,
 			Query: r.URL.Query().Encode(),
 		})
 		if err != nil {
-			http.Error(w, err.Error(), http.StatusBadRequest)
+			if identity.IsErrDuplicatedIdentity(err) {
+				err = accountmanagement.NewErrAccountManagementDuplicatedIdentity(err)
+			}
```
Comment (louischan-oursky) [inline pkg/auth/handler/webapp/sso_callback.go:62]: Why do we need to translate error here? Shouldn't this be done by AccountManagementService?

**Review by louischan-oursky — CHANGES_REQUESTED** [review]

**Review by louischan-oursky — CHANGES_REQUESTED** [review]
SCR-20241015-nluo
SCR-20241015-nlvk

This account has two identities, one email login ID, one google login. But it does not allow to remove the google login

**Comment by IniZio** [comment]
@louischan-oursky Updated to check other identity types as well 🙏

**Review by louischan-oursky — APPROVED** [review]

**Pull request merged**